client/client.go

```diff
@@ -136,77 +251,105 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 	return
 }
 
+func (client *Client) lockReconnect() (success bool) {
+	return atomic.CompareAndSwapUint32(&client.reconnectState, 0, 1)
+}
+
+// called by owner of reconnect state to tell that it has finished reconnecting
+func (client *Client) resetReconnectState() {
+	atomic.StoreUint32(&client.reconnectState, 0)
+}
+
 func (client *Client) reconnect(err error) error {
-	if client.conn != nil {
-		return nil
-	}
 
-	// TODO I doubt this error handling is right because it looks
-	// really complicated.
+	// not actioning on error if it's deemed Temporary
+	// we might want to take note of timestamp and eventually recycle this connection
+	// if it persists too long (even though classified as Temporary here)
 	if opErr, ok := err.(*net.OpError); ok {
-		if opErr.Timeout() {
-			client.err(err)
-		}
 		if opErr.Temporary() {
 			return nil
 		}
-
-		return err
 	}
 
-	if err != nil {
-		client.err(err)
+	ownReconnect := client.lockReconnect()
+
+	if !ownReconnect {
+		//Reconnect collision, this thread will exit and wait on next client.Lock() for other to complete reconnection
+		return nil
 	}
 
-	// If it is unexpected error and the connection wasn't
-	// closed by Gearmand, the client should close the conection
-	// and reconnect to job server.
-	client.Close()
+	defer client.resetReconnectState() // before releasing client lock we will reset reconnection state
 
-	conn, err := net.Dial(client.net, client.addr)
+	connVersion := client.loadConn().connVersion
+
+	client.Log(Error, fmt.Sprintf("Closing connection to %v due to error %v, will reconnect...", client.addr, err))
+	if closeErr := client.Close(); closeErr != nil {
+		client.Log(Warning, fmt.Sprintf("Non-fatal error %v, while closing connection to %v", closeErr, client.addr))
+	}
 
+	close(client.expected)
```

Comment:
I would still close these *after* changing them.

Otherwise I think we're ready to go.